### PR TITLE
Simplify switches on a statically known discriminant in MIR.

### PR DESCRIPTION
**Pull request by nullptrderef29**
This PR adds a new MIR pass: `SimplifyStaticSwitch`, which aims to simplify switches in MIR that match on a known discriminant, this mainly happens as a result of MIR inlining.

Addresses: [#111442 (comment)](https://github.com/rust-lang/rust/issues/111442#issuecomment-1546958336)

### CODE REVIEW

**Comment by scottmcm** [comment]
cc @cjgillot specifically, to comment how this should go along with #107009

File: compiler/rustc_mir_transform/src/simplify_static_switch.rs [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+    fn name(&self) -> &'static str {
+        "SimplifyStaticSwitch"
+    }
+
+    fn is_enabled(&self, sess: &Session) -> bool {
+        sess.mir_opt_level() >= 2
+    }
+
+    fn run_pass(&self, tcx: TyCtxt<'tcx>, body: &mut Body<'tcx>) {
+        debug!(body = %tcx.def_path_str(body.source.def_id()));
```
Comment (cjgillot) [inline compiler/rustc_mir_transform/src/simplify_static_switch.rs:49]:
```suggestion
        debug!(body = ?body.source.def_id());
```

`def_path_str` dooms the current compilation session.

File: compiler/rustc_mir_transform/src/simplify_static_switch.rs [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+#[instrument(level = "debug", skip(tcx, body), ret)]
+fn simplify_static_switch<'tcx>(tcx: TyCtxt<'tcx>, body: &mut Body<'tcx>) -> bool {
+    let mut new_targets = FxHashMap::default();
+    let predecessors: &IndexSlice<_, _> = body.basic_blocks.predecessors();
+    for (block, data) in traversal::preorder(body) {
+        if data.is_cleanup {
+            continue;
+        }
+
+        let predecessors: &[BasicBlock] = &predecessors[block];
```
Comment (cjgillot) [inline compiler/rustc_mir_transform/src/simplify_static_switch.rs:66]:
```suggestion
        let predecessors = &predecessors[block];
```

File: compiler/rustc_mir_transform/src/simplify_static_switch.rs [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+            remove_dead_blocks(tcx, body);
+        }
+    }
+}
+
+#[instrument(level = "debug", skip(tcx, body), ret)]
+fn simplify_static_switch<'tcx>(tcx: TyCtxt<'tcx>, body: &mut Body<'tcx>) -> bool {
+    let mut new_targets = FxHashMap::default();
+    let predecessors: &IndexSlice<_, _> = body.basic_blocks.predecessors();
+    for (block, data) in traversal::preorder(body) {
```
Comment (cjgillot) [inline compiler/rustc_mir_transform/src/simplify_static_switch.rs:61]: Why this particular order?

File: compiler/rustc_mir_transform/src/simplify_static_switch.rs [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+            if let TerminatorKind::Goto { .. } = data.terminator().kind {
+                for statement in data.statements.iter().rev() {
+                    match statement.kind {
+                        StatementKind::Assign(box (
+                            place,
+                            Rvalue::Aggregate(box AggregateKind::Adt(def_id, variant, ..), ..),
+                        )) if place == discr => {
+                            debug!(
+                                "{pred:?}: {place:?} = {}::{variant:?}; goto -> {block:?};",
+                                tcx.def_path_str(def_id)
```
Comment (cjgillot) [inline compiler/rustc_mir_transform/src/simplify_static_switch.rs:102]:
```suggestion
                                "{pred:?}: {place:?} = {def_id:?}::{variant:?}; goto -> {block:?};"
```

File: compiler/rustc_mir_transform/src/simplify_static_switch.rs [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+                StatementKind::Assign(box (place, Rvalue::Discriminant(discr)))
+                    if place == switched =>
+                {
+                    Some(discr)
+                },
+                _ => None,
+            }
+        }) else {
+            continue
+        };
```
Comment (cjgillot) [inline compiler/rustc_mir_transform/src/simplify_static_switch.rs:89]: You need to check that `place.local` is never borrowed, as we could have an indirect mutation between the assignment and the `SwitchInt`.
Likewise for `discr.local`.
You will also need to check that there is no write to a parent projection of `discr`. (For instance, discr is `a.b.c` and there is a write to `a.b` in between.)

File: compiler/rustc_mir_transform/src/simplify_static_switch.rs [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+                    match statement.kind {
+                        StatementKind::Assign(box (
+                            place,
+                            Rvalue::Aggregate(box AggregateKind::Adt(def_id, variant, ..), ..),
+                        )) if place == discr => {
+                            debug!(
+                                "{pred:?}: {place:?} = {}::{variant:?}; goto -> {block:?};",
+                                tcx.def_path_str(def_id)
+                            );
+                            let discr_ty = tcx.type_of(def_id).skip_binder();
```
Comment (cjgillot) [inline compiler/rustc_mir_transform/src/simplify_static_switch.rs:104]: You must take the substitutions into account here, otherwise you'll get the wrong type.

**Review by cjgillot — COMMENTED** [review]
We already have `SeparateConstSwitch` and `ConstProp`, which taken together do a similar transformation. What is the gain of this pass compared to the current state?

File: compiler/rustc_mir_transform/src/simplify_static_switch.rs (hunk 3 above)
Comment (JohnBobbo96) [reply compiler/rustc_mir_transform/src/simplify_static_switch.rs]: I thought it might be better than iterating normally but i have switched it back to just `iter_enumerated`.

File: compiler/rustc_mir_transform/src/simplify_static_switch.rs (hunk 5 above)
Comment (JohnBobbo96) [reply compiler/rustc_mir_transform/src/simplify_static_switch.rs:99]: i have added checks to make sure that `switched` and `discr` are both locals with no projections, and that they are not borrowed.

**Comment by JohnBobbo96** [comment]
> We already have `SeparateConstSwitch` and `ConstProp`, which taken together do a similar transformation. What is the gain of this pass compared to the current state?

I'd say the main advantage is that this pass sort of combines the effect of `SeparateConstSwitch` + `ConstProp` except it looks for aggregates instead of constants, which currently doesn't seem to be handled by `SeparateConstSwitch` + `ConstProp` alone, see this [godbolt link](https://rust.godbolt.org/z/qEdsfrfjM) (the block that switches on the discriminant of `_2` is still there, which this pass removes).

File: compiler/rustc_mir_transform/src/lib.rs [hunk 7]
```diff
@@ -130,7 +131,7 @@ pub fn provide(providers: &mut Providers) {
         mir_generator_witnesses: generator::mir_generator_witnesses,
         optimized_mir,
         is_mir_available,
-        is_ctfe_mir_available: |tcx, did| is_mir_available(tcx, did),
+        is_ctfe_mir_available: is_mir_available,
```
Comment (cjgillot) [inline compiler/rustc_mir_transform/src/lib.rs:134]: This change is unrelated. Could you submit it as a separate PR?

File: compiler/rustc_mir_transform/src/lib.rs [hunk 8]
```diff
@@ -532,7 +533,7 @@ fn run_runtime_cleanup_passes<'tcx>(tcx: TyCtxt<'tcx>, body: &mut Body<'tcx>) {
 }
 
 fn run_optimization_passes<'tcx>(tcx: TyCtxt<'tcx>, body: &mut Body<'tcx>) {
-    fn o1<T>(x: T) -> WithMinOptLevel<T> {
+    const fn o1<T>(x: T) -> WithMinOptLevel<T> {
```
Comment (cjgillot) [inline compiler/rustc_mir_transform/src/lib.rs:536]: Why is this useful?

File: compiler/rustc_mir_transform/src/simplify_static_switch.rs [hunk 9]
```diff
@@ -0,0 +1,165 @@
...
+///    \
+///     x = Foo::B(y); --------- Foo::B ---> ...
+/// ```
+pub struct SimplifyStaticSwitch;
+
+impl<'tcx> MirPass<'tcx> for SimplifyStaticSwitch {
+    fn name(&self) -> &'static str {
+        "SimplifyStaticSwitch"
+    }
+
```
Comment (cjgillot) [inline compiler/rustc_mir_transform/src/simplify_static_switch.rs:43]:
```suggestion
```
This is the default.

File: compiler/rustc_mir_transform/src/simplify_static_switch.rs [hunk 10]
```diff
@@ -0,0 +1,165 @@
...
+        }
+
+        let TerminatorKind::SwitchInt {
+            discr: Operand::Move(switched) | Operand::Copy(switched),
+            ref targets
+        } = data.terminator().kind else {
+            continue
+        };
+
+        let Some(switched) = switched.as_local() else {
```
Comment (cjgillot) [inline compiler/rustc_mir_transform/src/simplify_static_switch.rs:78]: Could you add a comment?

File: compiler/rustc_mir_transform/src/simplify_static_switch.rs [hunk 11]
```diff
@@ -0,0 +1,165 @@
...
+            continue
+        };
+
+        if borrowed_locals.contains(switched) {
```
Comment (cjgillot) [inline compiler/rustc_mir_transform/src/simplify_static_switch.rs:82]: Likewise.

File: compiler/rustc_mir_transform/src/simplify_static_switch.rs (hunk 5 above)
Comment (cjgillot) [reply compiler/rustc_mir_transform/src/simplify_static_switch.rs:99]: You also need to check that:
- `switched` is not overwritten between this assignment and the terminator;
- `discr` is not written between this assignment and the beginning of the block.

File: compiler/rustc_mir_transform/src/simplify_static_switch.rs [hunk 12]
```diff
@@ -0,0 +1,165 @@
...
+                        Rvalue::Aggregate(box AggregateKind::Adt(def_id, variant, substs, ..), ..),
+                    )) if place == discr => {
+                        debug!("{pred:?}: {place:?} = {def_id:?}::{variant:?}; goto -> {block:?};");
+                        let discr_ty = tcx.type_of(def_id).subst(tcx, substs);
+                        if let Some(discr) = discr_ty.discriminant_for_variant(tcx, variant) {
+                            let target = targets.target_for_value(discr.val);
+                            new_targets
+                                .entry((place, block))
+                                .or_insert_with(Vec::new)
+                                .push((pred, target));
```
Comment (cjgillot) [inline compiler/rustc_mir_transform/src/simplify_static_switch.rs:120]: There is a subtle invariant here: that the `(pred, target)` edge only appears once in the cfg. This is fulfilled in your implementation as `pred` ends with a `goto`, but this deserves to be commented.

File: compiler/rustc_mir_transform/src/simplify_static_switch.rs [hunk 13]
```diff
@@ -0,0 +1,165 @@
...
+        let block = &basic_blocks[block];
+        let statements = if block.statements.len() > 1 {
+            block
+                .statements
+                .iter()
+                .filter_map(|statement| match statement.kind {
+                    StatementKind::Assign(box (.., Rvalue::Discriminant(discr)))
+                        if place == discr =>
+                    {
+                        None
```
Comment (cjgillot) [inline compiler/rustc_mir_transform/src/simplify_static_switch.rs:144]: Why remove this statement? The LHS (`switched`) may still be used, even if we remove the `SwitchInt`.

File: compiler/rustc_mir_transform/src/simplify_static_switch.rs [hunk 14]
```diff
@@ -0,0 +1,165 @@
...
+            }
+        }
+    }
+
+    if new_targets.is_empty() {
+        return false;
+    }
+
+    let basic_blocks = body.basic_blocks.as_mut();
+    for ((place, block), new_targets) in new_targets {
```
Comment (cjgillot) [inline compiler/rustc_mir_transform/src/simplify_static_switch.rs:134]: This is what you call `discr` earlier, could you name it so here?

File: compiler/rustc_mir_transform/src/simplify_static_switch.rs [hunk 15]
```diff
@@ -0,0 +1,165 @@
...
+                        None
+                    }
+                    _ => Some(statement.clone()),
+                })
+                .collect()
+        } else {
+            vec![]
+        };
+
+        for (block, target) in new_targets {
```
Comment (cjgillot) [inline compiler/rustc_mir_transform/src/simplify_static_switch.rs:153]: `block` here is `pred`, please call it so.

File: compiler/rustc_mir_transform/src/simplify_static_switch.rs [hunk 16]
```diff
@@ -0,0 +1,165 @@
...
+            let data = &mut basic_blocks[block];
+            let old_goto = &mut data.terminator.as_mut().unwrap().kind;
```
Comment (cjgillot) [inline compiler/rustc_mir_transform/src/simplify_static_switch.rs:155]:
```suggestion
            let old_goto = &mut data.terminator_mut().kind;
```

File: compiler/rustc_mir_transform/src/lib.rs [hunk 17]
```diff
@@ -548,6 +549,8 @@ fn run_optimization_passes<'tcx>(tcx: TyCtxt<'tcx>, body: &mut Body<'tcx>) {
             &uninhabited_enum_branching::UninhabitedEnumBranching,
             &o1(simplify::SimplifyCfg::AfterUninhabitedEnumBranching),
             &inline::Inline,
+            // Remove switches on a statically known discriminant, which can happen as a result of inlining.
+            &simplify_static_switch::SimplifyStaticSwitch,
```
Comment (cjgillot) [inline compiler/rustc_mir_transform/src/lib.rs:553]: This pass clones statements. As we have SSA-based passes, and we don't try to recover SSAness, it should be run much later.

File: compiler/rustc_mir_transform/src/simplify_static_switch.rs [hunk 18]
```diff
@@ -0,0 +1,165 @@
...
+        if borrowed_locals.contains(switched) {
+            continue;
+        }
+
+        let Some(discr) = data.statements.iter().rev().find_map(|statement| {
+            match statement.kind {
+                StatementKind::Assign(box (place, Rvalue::Discriminant(discr)))
+                    if place.as_local() == Some(switched)
+                        && let Some(local) = discr.as_local()
+                        && !borrowed_locals.contains(local) =>
```
Comment (cjgillot) [inline compiler/rustc_mir_transform/src/simplify_static_switch.rs:91]: Those checks should happen outside of the `find_map`, in cases there are several assignments to `switched`.

File: compiler/rustc_mir_transform/src/simplify_static_switch.rs (hunk 5 above)
Comment (JohnBobbo96) [reply compiler/rustc_mir_transform/src/simplify_static_switch.rs:110]: i added a function `is_local_mutated` to make sure that `switched` and `discr` do not get overwritten.

File: compiler/rustc_mir_transform/src/lib.rs (hunk 7 above)
Comment (JohnBobbo96) [reply compiler/rustc_mir_transform/src/lib.rs]: i removed this for now, i'll open a PR soon.

File: compiler/rustc_mir_transform/src/lib.rs (hunk 8 above)
Comment (JohnBobbo96) [reply compiler/rustc_mir_transform/src/lib.rs]: i also removed this.

File: compiler/rustc_mir_transform/src/simplify_static_switch.rs (hunk 12 above)
Comment (JohnBobbo96) [reply compiler/rustc_mir_transform/src/simplify_static_switch.rs]: i added a comment mentioning this.

File: compiler/rustc_mir_transform/src/simplify_static_switch.rs (hunk 13 above)
Comment (JohnBobbo96) [reply compiler/rustc_mir_transform/src/simplify_static_switch.rs]: now all of the statements are cloned instead of ignoring this statement.

File: compiler/rustc_mir_transform/src/lib.rs (hunk 17 above)
Comment (JohnBobbo96) [reply compiler/rustc_mir_transform/src/lib.rs]: i have moved the pass to right before `SeparateConstSwitch`.

File: compiler/rustc_mir_transform/src/simplify_static_switch.rs (hunk 18 above)
Comment (JohnBobbo96) [reply compiler/rustc_mir_transform/src/simplify_static_switch.rs]: i also moved these checks to outside of the `find_map`.

File: compiler/rustc_mir_transform/src/simplify_static_switch.rs [hunk 19]
```diff
@@ -0,0 +1,203 @@
...
+        };
+
+        if borrowed_locals.contains(discr) {
+            continue;
+        }
+
+        // We need to make sure that `switched` is not overwritten before the switch,
+        // and also that `discr` is not overwritten before the `Rvalue::Discriminant`.
+        if is_local_mutated(body, block, discr, index, 0..index)
+            || is_local_mutated(body, block, switched, index, index..data.statements.len())
```
Comment (cjgillot) [inline compiler/rustc_mir_transform/src/simplify_static_switch.rs:119]: Nit: can we just pass a range to this function?

File: compiler/rustc_mir_transform/src/simplify_static_switch.rs [hunk 20]
```diff
@@ -0,0 +1,203 @@
...
+        }
+    }
+}
+
+#[instrument(level = "debug", skip(tcx, body), ret)]
+fn simplify_static_switch<'tcx>(tcx: TyCtxt<'tcx>, body: &mut Body<'tcx>) -> bool {
+    let predecessors = body.basic_blocks.predecessors();
+    let borrowed_locals = borrowed_locals(body);
+    // Should this be an `IndexVec` instead?
+    let mut new_targets = FxHashMap::default();
```
Comment (cjgillot) [inline compiler/rustc_mir_transform/src/simplify_static_switch.rs:63]: A simple `Vec` should be enough, as you don't use it for indexing.

File: compiler/rustc_mir_transform/src/simplify_static_switch.rs [hunk 21]
```diff
@@ -0,0 +1,203 @@
...
+            let TerminatorKind::Goto { .. } = data.terminator().kind else {
+                continue
+            };
+
+            for statement in data.statements.iter().rev() {
+                match statement.kind {
+                    StatementKind::Assign(box (
+                        place,
+                        Rvalue::Aggregate(box AggregateKind::Adt(def_id, variant, substs, ..), ..),
+                    )) if place.as_local() == Some(discr) => {
```
Comment (cjgillot) [inline compiler/rustc_mir_transform/src/simplify_static_switch.rs:135]: You need to check if `discr` is mutated between this assignment and the end of block.

File: compiler/rustc_mir_transform/src/simplify_static_switch.rs [hunk 22]
```diff
@@ -0,0 +1,203 @@
...
+        }
+
+        for &pred in predecessors {
+            let data = &body.basic_blocks[pred];
+            let TerminatorKind::Goto { .. } = data.terminator().kind else {
+                continue
+            };
+
+            for statement in data.statements.iter().rev() {
+                match statement.kind {
```
Comment (cjgillot) [inline compiler/rustc_mir_transform/src/simplify_static_switch.rs:131]: Nit: this can be an if-let.

File: compiler/rustc_mir_transform/src/simplify_static_switch.rs [hunk 23]
```diff
@@ -0,0 +1,203 @@
...
+
+    if new_targets.is_empty() {
+        return false;
+    }
+
+    let basic_blocks = body.basic_blocks.as_mut();
+    for (block, new_targets) in new_targets {
+        let statements: Vec<_> = basic_blocks[block].statements.iter().cloned().collect();
+        for (pred, target) in new_targets {
+            let pred = &mut basic_blocks[pred];
```
Comment (cjgillot) [inline compiler/rustc_mir_transform/src/simplify_static_switch.rs:157]: Please document why there is only a single `pred -> block` edge in the whole CFG, and that no earlier threading has overwritten it.

File: compiler/rustc_mir_transform/src/simplify_static_switch.rs [hunk 24]
```diff
@@ -0,0 +1,203 @@
...
+            let old_goto = &mut pred.terminator_mut().kind;
```
Comment (cjgillot) [inline compiler/rustc_mir_transform/src/simplify_static_switch.rs:158]: We should assert that `old_goto` is indeed `goto block`, to ensure we have not overwritten it in the mean time.

File: compiler/rustc_mir_transform/src/simplify_static_switch.rs [hunk 25]
```diff
@@ -0,0 +1,203 @@
...
+        }
+    }
+
+    if new_targets.is_empty() {
+        return false;
+    }
+
+    let basic_blocks = body.basic_blocks.as_mut();
+    for (block, new_targets) in new_targets {
+        let statements: Vec<_> = basic_blocks[block].statements.iter().cloned().collect();
```
Comment (cjgillot) [inline compiler/rustc_mir_transform/src/simplify_static_switch.rs:155]: Could we use `IndexSlice::pick2_mut` to avoid having to clone statements twice?

File: compiler/rustc_mir_transform/src/simplify_static_switch.rs (hunk 23 above)
Comment (cjgillot) [reply compiler/rustc_mir_transform/src/simplify_static_switch.rs:157]: It may also be useful to document why `pred != block` and assert it.

File: tests/mir-opt/simplify_static_switch.rs [hunk 26]
```diff
@@ -0,0 +1,29 @@
+// unit-test: SimplifyStaticSwitch
```
Comment (cjgillot) [inline tests/mir-opt/simplify_static_switch.rs:1]: Could you add more tests?
With all the extra checks I made you add + weird cfgs if possible.
See https://github.com/rust-lang/rust/pull/107009 for a few examples.

File: compiler/rustc_mir_transform/src/simplify_static_switch.rs [hunk 27]
```diff
@@ -0,0 +1,203 @@
...
+        for (pred, target) in new_targets {
+            let pred = &mut basic_blocks[pred];
+            let old_goto = &mut pred.terminator_mut().kind;
+            let new_goto = TerminatorKind::Goto { target };
+            debug!("{block:?}: old = {old_goto:?}; new = {new_goto:?};");
+
+            // Something to be noted is that, this creates an edge from `pred -> target`
+            // which will only appear once in the CFG.
+            *old_goto = new_goto;
```
Comment (cjgillot) [inline compiler/rustc_mir_transform/src/simplify_static_switch.rs:164]: Should we prevent threading through a loop header, in order to avoid creating irreducible CFGs?

**Review by cjgillot — COMMENTED** [review]
Thanks! This is starting to look good.
Could you add comments explaining the pattern you are looking for in this opt, and why it is sound?

File: compiler/rustc_mir_transform/src/simplify_static_switch.rs (hunk 21 above)
Comment (JohnBobbo96) [reply compiler/rustc_mir_transform/src/simplify_static_switch.rs]: I have added a check for this.

File: compiler/rustc_mir_transform/src/simplify_static_switch.rs (hunk 24 above)
Comment (JohnBobbo96) [reply compiler/rustc_mir_transform/src/simplify_static_switch.rs]: I have added an assert to check this, as well as not allowing two `static_switches` to have the same `pred`, so the terminator should not get overwritten.

File: compiler/rustc_mir_transform/src/simplify_static_switch.rs (hunk 25 above)
Comment (JohnBobbo96) [reply compiler/rustc_mir_transform/src/simplify_static_switch.rs]: This is cool, i also added this.

File: tests/mir-opt/simplify_static_switch.rs (hunk 26 above)
Comment (JohnBobbo96) [reply tests/mir-opt/simplify_static_switch.rs:1]: I have added a few more tests, including checks for mutation/borrowing.

File: compiler/rustc_mir_transform/src/simplify_static_switch.rs (hunk 27 above)
Comment (JohnBobbo96) [reply compiler/rustc_mir_transform/src/simplify_static_switch.rs]: I also added a check for this.

**Comment by JohnBobbo96** [comment]
> Thanks! This is starting to look good. Could you add comments explaining the pattern you are looking for in this opt, and why it is sound?

Thank you for reviewing this! I have also added more to the documentation explaining why believe it is sound.

File: compiler/rustc_mir_transform/src/simplify_static_switch.rs [hunk 28]
```diff
@@ -0,0 +1,320 @@
...
+
+        let terminator = data.terminator();
+        let TerminatorKind::SwitchInt {
+            discr: Operand::Copy(place) | Operand::Move(place),
+            targets
+        } = &terminator.kind else {
+            continue
+        };
+
+        if place.local != switched {
```
Comment (cjgillot) [inline compiler/rustc_mir_transform/src/simplify_static_switch.rs:176]: `local` has necessarily an integer type. Should we assert that `place.projection` is empty?

File: compiler/rustc_mir_transform/src/simplify_static_switch.rs [hunk 29]
```diff
@@ -0,0 +1,320 @@
...
+            continue;
+        }
+
+        let mut finder = MutatedLocalFinder { local: discr, mutated: false };
```
Comment (cjgillot) [inline compiler/rustc_mir_transform/src/simplify_static_switch.rs:180]: This should be created inside the loop, to ensure we don't forget to reset `mutated`.

File: compiler/rustc_mir_transform/src/simplify_static_switch.rs [hunk 30]
```diff
@@ -0,0 +1,320 @@
...
+                continue
+            };
+
+            for (statement_index, statement) in data.statements.iter().enumerate().rev() {
+                match statement.kind {
+                    StatementKind::SetDiscriminant { box place, variant_index: variant }
+                    | StatementKind::Assign(box (
+                        place,
+                        Rvalue::Aggregate(box AggregateKind::Adt(_, variant, ..), ..),
+                    )) if place.local == discr => {
```
Comment (cjgillot) [inline compiler/rustc_mir_transform/src/simplify_static_switch.rs:194]: Assert `place.projection` is empty here too.

File: compiler/rustc_mir_transform/src/simplify_static_switch.rs [hunk 31]
```diff
@@ -0,0 +1,320 @@
...
+                    _ if finder.mutated => {
+                        debug!("The discriminant: {discr:?} was mutated in predecessor: {pred:?}");
+                        // Note that the discriminant could have been mutated in one predecessor
+                        // but not the others, in which case only the predecessors which did not mutate
+                        // the discriminant can be optimized.
+                        finder.mutated = false;
+
+                        continue 'preds;
+                    }
+                    _ => finder.visit_statement(statement, Location { block, statement_index }),
```
Comment (cjgillot) [inline compiler/rustc_mir_transform/src/simplify_static_switch.rs:247]: Could we create `finder`, visit, and check `mutated` in this arm, instead of the other arms?

File: compiler/rustc_mir_transform/src/simplify_static_switch.rs [hunk 32]
```diff
@@ -0,0 +1,320 @@
...
+#[instrument(level = "debug", skip_all, ret)]
+fn simplify_static_switches<'tcx>(
+    tcx: TyCtxt<'tcx>,
+    body: &mut Body<'tcx>,
+    ssa_locals: &SsaLocals,
+) -> bool {
+    let dominators = body.basic_blocks.dominators();
+    let predecessors = body.basic_blocks.predecessors();
+    let mut discriminants = FxHashMap::default();
+    let mut static_switches = FxHashMap::default();
```
Comment (cjgillot) [inline compiler/rustc_mir_transform/src/simplify_static_switch.rs:87]: Could you document what is store in each map?

File: compiler/rustc_mir_transform/src/simplify_static_switch.rs [hunk 33]
```diff
@@ -0,0 +1,320 @@
...
+            },
+        );
+    }
+
+    if discriminants.is_empty() {
+        debug!("No SSA locals were assigned a discriminant");
+        return false;
+    }
+
+    for (switched, Discriminant { discr, block, exclude }) in discriminants {
```
Comment (cjgillot) [inline compiler/rustc_mir_transform/src/simplify_static_switch.rs:147]: `discriminants` is a hashmap. Iteration order is unstable. Should `discriminants` be a Vec or pairs?

File: compiler/rustc_mir_transform/src/simplify_static_switch.rs [hunk 34]
```diff
@@ -0,0 +1,320 @@
...
+    }
+
+    if static_switches.is_empty() {
+        debug!("No static switches were found in the current body");
+        return false;
+    }
+
+    let basic_blocks = body.basic_blocks.as_mut();
+    let num_switches: usize = static_switches.iter().map(|(_, switches)| switches.len()).sum();
+    for (block, static_switches) in static_switches {
```
Comment (cjgillot) [inline compiler/rustc_mir_transform/src/simplify_static_switch.rs:260]: Likewise, `static_switches` is a hashmap, iteration order is unstable.

File: compiler/rustc_mir_transform/src/simplify_static_switch.rs [hunk 35]
```diff
@@ -0,0 +1,320 @@
...
+                                            tcx.arena.alloc_from_iter(
+                                                static_switches.iter().copied().chain([
+                                                    StaticSwitch { pred, target, exclude },
+                                                ]),
+                                            );
+                                    }
+                                })
+                                .or_insert_with(|| {
+                                    tcx.arena.alloc([StaticSwitch { pred, target, exclude }])
+                                });
```
Comment (cjgillot) [inline compiler/rustc_mir_transform/src/simplify_static_switch.rs:233]: I don't understand how the worklist is stored and why.
Could you add an explanation?
Why isn't a vector of edges `(pred, block, target)`, meaning "the pred->block edge becomes pred->target" enough?

File: compiler/rustc_mir_transform/src/simplify_static_switch.rs [hunk 36]
```diff
@@ -0,0 +1,320 @@
...
+            debug!("{block:?}: Removing static switch: {switch:?}");
+
+            // We use the SSA, to destroy the SSA.
+            let data = {
+                let (block, pred) = basic_blocks.pick2_mut(block, switch.pred);
+                match switch.exclude {
+                    Some(exclude) => {
+                        pred.statements.extend(block.statements.iter().enumerate().filter_map(
+                            |(index, statement)| {
+                                if index == exclude { None } else { Some(statement.clone()) }
```
Comment (cjgillot) [inline compiler/rustc_mir_transform/src/simplify_static_switch.rs:271]: That's not necessary. We can run `remove_unused_definitions` together with `remove_dead_blocks`.

File: compiler/rustc_mir_transform/src/simplify_static_switch.rs [hunk 37]
```diff
@@ -0,0 +1,320 @@
...
+struct MutatedLocalFinder {
+    local: Local,
+    mutated: bool,
+}
+
+impl<'tcx> Visitor<'tcx> for MutatedLocalFinder {
+    fn visit_place(&mut self, place: &Place<'tcx>, context: PlaceContext, _: Location) {
+        if self.local == place.local && let PlaceContext::MutatingUse(..) = context {
+            self.mutated = true;
+        }
```
Comment (cjgillot) [inline compiler/rustc_mir_transform/src/simplify_static_switch.rs:318]: Should we detect storage statements too?

File: tests/mir-opt/separate_const_switch.identity.SeparateConstSwitch.diff [hunk 38]
```diff
@@ -51,51 +51,44 @@
           StorageLive(_10);
           StorageLive(_11);
           _9 = discriminant(_1);
-          switchInt(move _9) -> [0: bb7, 1: bb5, otherwise: bb6];
+          switchInt(move _9) -> [0: bb5, 1: bb3, otherwise: bb4];
```
Comment (cjgillot) [inline tests/mir-opt/separate_const_switch.identity.SeparateConstSwitch.diff:54]: Should we make this a unit test for `SeparateConstSwitch`?

File: compiler/rustc_mir_transform/src/simplify_static_switch.rs [hunk 39]
```diff
@@ -0,0 +1,320 @@
...
+
+    let basic_blocks = body.basic_blocks.as_mut();
+    let num_switches: usize = static_switches.iter().map(|(_, switches)| switches.len()).sum();
+    for (block, static_switches) in static_switches {
+        for switch in static_switches {
+            debug!("{block:?}: Removing static switch: {switch:?}");
+
+            // We use the SSA, to destroy the SSA.
+            let data = {
+                let (block, pred) = basic_blocks.pick2_mut(block, switch.pred);
```
Comment (cjgillot) [inline compiler/rustc_mir_transform/src/simplify_static_switch.rs:266]: Could you add a comment explaining why `block != switch.pred`?

File: compiler/rustc_mir_transform/src/simplify_static_switch.rs [hunk 40]
```diff
@@ -0,0 +1,320 @@
...
+        }
+    }
+
+    if static_switches.is_empty() {
+        debug!("No static switches were found in the current body");
+        return false;
+    }
+
+    let basic_blocks = body.basic_blocks.as_mut();
+    let num_switches: usize = static_switches.iter().map(|(_, switches)| switches.len()).sum();
```
Comment (cjgillot) [inline compiler/rustc_mir_transform/src/simplify_static_switch.rs:259]: Counting the number of opts just for tracing is not useful.

**Comment by JohnCSimon** [comment]
@JohnBobbo96 

ping from triage - can you post your status on this PR? There hasn't been an update in a few months and there are merge conflicts Thanks!


FYI: when a PR is ready for review, send a message containing 
`@rustbot ready` to switch to `S-waiting-on-review` so the PR is in the reviewer's backlog.

**Comment by JohnCSimon** [comment]
@JohnBobbo96 


Ping from triage: I'm closing this due to inactivity, Please reopen when you are ready to continue with this. 
Note: if you are going to continue please open the PR **BEFORE** you push to it, else you won't be able to reopen - this is a quirk of github. 
Thanks for your contribution.

@rustbot label: +S-inactive
